corehq/apps/userreports/tests/test_report_aggregation.py

```diff
@@ -739,3 +757,89 @@ def test_with_prefilter(self):
                 ]
             ]]
         )
+
+    def test_aggregate_date(self):
+        report_config = self._create_report(
+            aggregation_columns=[
+                'indicator_col_id_state',
+                'month',
+            ],
+            columns=[
+                {
+                    'type': 'field',
+                    'display': 'report_column_display_state',
+                    'field': 'indicator_col_id_state',
+                    'column_id': 'report_column_col_id_state',
+                    'aggregation': 'simple'
+                },
+                {
+                    'type': 'aggregate_date',
+                    'display': 'month',
+                    'field': 'date',
+                    'column_id': 'month',
+                    'aggregation': 'simple',
+                    'format': '%Y-%m'
+                },
+                {
+                    'type': 'field',
+                    'display': 'report_column_display_number',
+                    'field': 'indicator_col_id_number',
+                    'column_id': 'report_column_col_id_number',
+                    'aggregation': 'sum'
+                }
+            ],
+            filters=None,
+        )
+        view = self._create_view(report_config)
+        self.assertEqual(
+            view.export_table,
+            [['foo',
+              [['report_column_display_state', 'month', 'report_column_display_number'],
+               ['MA', '2018-01', 6],
+               ['MA', '2018-02', 3],
+               ['TN', '2017-01', 1]]]]
+        )
+
+    def test_conditional_aggregation(self):
+        report_config = self._create_report(
+            aggregation_columns=[
+                'indicator_col_id_state',
+                'age_range',
+            ],
+            columns=[
+                {
+                    'type': 'field',
+                    'display': 'state',
+                    'field': 'indicator_col_id_state',
+                    'column_id': 'state',
+                    'aggregation': 'simple'
+                },
+                {
+                    'type': 'conditional_aggregation',
+                    'display': 'age_range',
+                    'field': 'age_at_registration',
+                    'column_id': 'age_range',
+                    'whens': {
+                        "age_at_registration between 0 and 6": "0-6",
+                        "age_at_registration between 7 and 12": "7-12",
+                    },
+                    'else_': '13+'
+                },
```

Comment:
This test shows the functionality that's exposed.  The constrained version I'm considering would look something like this:
```JSON
{
    "type": "aggregate_range",
    "display": "age_range",
    "field": "age_at_registration",
    "column_id": "age_range",
    "ranges": {
        "0-6": [0, 6],
        "7-12": [7, 12],
    },
    "default": "13+"
}
```
That is, the user provides names, start and end points, and a default value, and that is transformed to the above.

To support similar use-cases, we'd need to write similar wrappers.

Reply:
https://erikbern.com/2018/08/30/i-dont-want-to-learn-your-garbage-query-language.html

Reply:
hahaha this is great.

> There should be a 30 year moratorium on inventing new query languages.

I've been thinking about this a lot (specifically in the context of UCR - which I think is a wonderful example of the pain of a custom query language). The thing I haven't figured out is how to get the benefits/reusability of a non-custom language without at least one of the following drawbacks:

- potentially exposing security holes by passing through raw queries
- doing a hell of a lot of work to build a giant custom parser
- supporting a bastardized subset of the query language that can be just as confusing as a custom one
- ~impossible to build a non-power-user UI on top of

the author suggests they just want to "download their data to sql" which is fine for the author but not really tenable in a SaaS product.

anyway, nothing actionable in this comment, but it resonates and also I don't quite know what the good alternative is.

Reply:
I should add - generally I'm very much in favor of moving away from a custom query language unless we're willing to build a UI on top of it. The obvious choices for UCR would be to use `xpath` (better for our non-power users) or `sql` (better for our power-users and easier for our dev team). I also like the idea of transitioning more to SQL, though don't have my head wrapped around what the security precautions/rules we need to make are if we start to go down that route...

Reply:
I had actually started something like this feature a while ago, but never quite got the interface that I was happy with. Reading that also really resonated with me and I think that putting SQL like this in is great. 

I don't believe this is a security hole as it's being escaped by sqlalchemy and I also don't believe it's anymore of a performance concern than UCR already is because it's not trying to query these columns, just performing calculations on the rows that would have already been returned

Reply:
Got any thoughts on how/whether to document this?  Just say it's a SQL case expression and leave it at that?  Seems like it'd be best explained with examples.

Reply:
to start you could add it to the readme with a link to [an example here](https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/userreports/examples/examples.md#data-source-indicators)